src/trading/CompleteSets.sol

```diff
@@ -0,0 +1,73 @@
+pragma solidity ^0.4.13;
+
+import 'ROOT/trading/ICompleteSets.sol';
+import 'ROOT/Controlled.sol';
+import 'ROOT/libraries/ReentrancyGuard.sol';
+import 'ROOT/libraries/math/SafeMathUint256.sol';
+import 'ROOT/libraries/token/ERC20.sol';
+import 'ROOT/extensions/MarketFeeCalculator.sol';
+import 'ROOT/reporting/IMarket.sol';
+import 'ROOT/reporting/IReportingWindow.sol';
+import 'ROOT/trading/IOrders.sol';
+
+
+contract CompleteSets is Controlled, ReentrancyGuard, ICompleteSets {
+    using SafeMathUint256 for uint256;
+
+    /**
+     * Buys `_amount` shares of every outcome in the specified market.
+    **/
+    function publicBuyCompleteSets(IMarket _market, uint256 _amount) external onlyInGoodTimes nonReentrant returns (bool) {
+        return this.buyCompleteSets(msg.sender, _market, _amount);
+    }
+
+    function buyCompleteSets(address _sender, IMarket _market, uint256 _amount) external onlyWhitelistedCallers returns (bool) {
+        require(_sender != address(0));
+        require(_market != IMarket(0));
+
+        uint8 _numOutcomes = _market.getNumberOfOutcomes();
+        ERC20 _denominationToken = _market.getDenominationToken();
+
+        require(_denominationToken.transferFrom(_sender, _market, _amount));
+        for (uint8 _outcome = 0; _outcome < _numOutcomes; ++_outcome) {
+            _market.getShareToken(_outcome).createShares(_sender, _amount);
+        }
+
+        IOrders(controller.lookup("Orders")).buyCompleteSetsLog(_sender, _market, _amount, _numOutcomes);
+        return true;
+    }
+
+    function publicSellCompleteSets(IMarket _market, uint256 _amount) external onlyInGoodTimes nonReentrant returns (bool) {
+        return this.sellCompleteSets(msg.sender, _market, _amount);
+    }
+
+    function sellCompleteSets(address _sender, IMarket _market, uint256 _amount) external onlyWhitelistedCallers returns (bool) {
+        require(_sender != address(0));
+        require(_market != IMarket(0));
+
+        uint8 _numOutcomes = _market.getNumberOfOutcomes();
+        ERC20 _denominationToken = _market.getDenominationToken();
+        uint256 _marketCreatorFeeRate = _market.getMarketCreatorSettlementFeeInAttoethPerEth();
+        uint256 _marketCreatorFee = _amount.mul(_marketCreatorFeeRate).div(10**18);
```

Comment:
Small nit but I prefer `1 ether` instead of `10**18`

Reply:
Fixed.